### PR TITLE
Failure healing policies

### PR DESCRIPTION
**Pull request by zofiaOS**

### CODE REVIEW

File: plugins/riemann-controller/riemann_controller/config.py [hunk 1]
```diff
@@ -27,11 +27,13 @@ def create(ctx, policy_types, policy_triggers, groups, config_template):
                 'group': group_name,
                 'policy': policy_name,
                 'policy_type': policy['type'],
+                'policy_properties': policy['properties'],
```
Comment (dankilman) [inline plugins/riemann-controller/riemann_controller/config.py]: Policy properties are already injected as the top level `template_vars`, there is no need to add them again into the metadata

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 2]
```diff
@@ -23,28 +30,43 @@
...
+           :deployment_id "{{ctx.deployment.id}}"
+           :group         "{{stream.metadata.group}}"
+           :policy        "{{stream.metadata.policy}}"}
+      process-policy-triggers      (process-policy-triggers-stream
+                                     ctx deployment-processing-queue)
+      publish-policy-event         (publish-policy-event* ctx)
+      publish-policy-error         (publish-policy-error* ctx)
+      publish-log                  (publish-log* ctx)
+      restraints                   (get-workflow-restraints
+                                     (parse-boolean "{{stream.metadata.policy_properties.is_node_started_before_workflow}}")
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: Leave some clojure comments about unspecified values being handled correctly

File: plugins/riemann-controller/riemann_controller/config_constants.py [hunk 3]
```diff
@@ -1,3 +1,23 @@
...
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+
 
 class Constants:
+    """ Constants that can be used in policies """
     PERIODICAL_EXPIRATION_INTERVAL = 5  # in seconds
+    TRIGGERING_STATE = "triggering_state"
+    STABLE_STATE = "ok"
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/config_constants.py]: move to clojure or delete 20, 21

File: plugins/riemann-controller/riemann_controller/config_constants.py [hunk 4]
```diff
@@ -1,3 +1,23 @@
...
+    HEART_BEAT_FAILURE = "heart-beat-failure"
+    MIN_INTERVAL_BETWEEN_WORKFLOWS = 5
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/config_constants.py]: delete

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 5]
```diff
@@ -5,7 +5,8 @@
 {% for trigger_name, trigger in policy_triggers.iteritems()  %}
 
 (register-policy-trigger "{{ctx.deployment.id}}" "{{trigger_name}}"
-{{trigger.source}})
+{{trigger.source}}
+)
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: add a comment

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 6]
```diff
@@ -14,6 +15,12 @@
         (processing-queue "{{ctx.deployment.id}}" {:queue-size     1000
                                                    :core-pool-size 1
                                                    :max-pool-size  4})]
+(def node-instances-stream {
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: make it optional - sometimes we want to define policy for a group of nodes. two options:
1. node instances
2. all that are part of the members section

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 7]
```diff
@@ -23,28 +30,43 @@
...
+                                     ctx deployment-processing-queue)
+      publish-policy-event         (publish-policy-event* ctx)
+      publish-policy-error         (publish-policy-error* ctx)
+      publish-log                  (publish-log* ctx)
+      restraints                   (get-workflow-restraints
+                                     (parse-boolean "{{stream.metadata.policy_properties.is_node_started_before_workflow}}")
+                                     (or {{stream.metadata.policy_properties.interval_between_workflows}} 300))
+      check-restraints-and-process (fn [restraints]
+        (check-restraints-and-process* ctx deployment-processing-queue restraints))
+      downstream
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: make the names: downstream, restraints so it is more specific

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 8]
```diff
@@ -23,28 +30,43 @@
...
+        (autohealing/downstream* index
+                                 (check-restraints-and-process restraints)
+                                 "{{constants.STABLE_STATE}}"
+                                 "{{constants.TRIGGERING_STATE}}")
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: move  downstream to the policies

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 9]
```diff
@@ -23,28 +30,43 @@
...
+      stream_id "{{stream.metadata.group}}.{{stream.metadata.policy}}"]
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: should have more private name

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 10]
```diff
@@ -23,28 +30,43 @@
...
+    (fn [event]
+      (def node-instances-stream
+        (if (contains? (node-instances-stream stream_id) (:node_id event))
+          node-instances-stream
+          (assoc node-instances-stream
+                 stream_id
+                 (assoc (node-instances-stream stream_id)
+                        (:node_id event)
+                        (sdo
+                          {{stream.data}}
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: add a comment about a new line

File: plugins/riemann-controller/riemann_controller/resources/manager.config [hunk 11]
```diff
@@ -408,17 +398,71 @@
...
-                                     :trigger policy-trigger-name
-                                     :trigger-parameters parsed-parameters}
+              ctx                   (assoc ctx :node-id (:node_id event)
+                                               :trigger policy-trigger-name
+                                               :trigger-parameters parsed-parameters)
               trigger-event         (assoc event :ctx ctx
                                                  :trigger trigger)]
           (deployment-processing-queue trigger-event))))))
 
+(defn check-restraints
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/manager.config]: rename it

File: plugins/riemann-controller/riemann_controller/resources/manager.config [hunk 12]
```diff
@@ -408,17 +398,71 @@
...
+    (if (check-restraints ctx restraints)
+      ((process-policy-triggers-stream ctx deployment-processing-queue) event))))
+
+(defn check-restraints-and-process* [ctx deployment-processing-queue restraints]
+  (fn [event]
+    (let [node-id       (:node_id event)
+          ctx           (assoc ctx :node-id (if node-id node-id (:failing_node event)))
+          trigger-event {:ctx ctx
+                         :trigger (check-restraints-and-process** deployment-processing-queue event restraints)}]
+      (deployment-processing-queue trigger-event))))
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/manager.config]: add an error  message

File: plugins/riemann-controller/riemann_controller/resources/manager.config [hunk 13]
```diff
@@ -408,17 +398,71 @@
...
+        manager-rest-port     (Integer/parseInt raw-manager-rest-port)
+        base-uri              (str "http://" manager-ip ":" manager-rest-port)
+        node-endpoint         (str "/node-instances/" node-id)
+        node-resource-uri     (str base-uri node-endpoint)
+        get-node              (fn [] (clj-http.client/get node-resource-uri {:accept :json}))
+        get-state             (fn [node-response]
+                                (:state (cheshire.core/parse-string (:body node-response) true)))]
+    (try
+      (= (get-state (get-node)) "started")
+      (catch Exception e false))))
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/manager.config]: maybe check a specific http error code - 404

File: resources/rest-service/cloudify/policies/ewma_stabilized.clj [hunk 14]
```diff
@@ -0,0 +1,7 @@
+(where (and (service #"{{service}}")
+            (not (expired? event)))
+    (ewma-timeless {{ewma_timeless_r}}
+      (where ((autohealing/inequality "{{upper_bound}}") metric {{threshold}})
```
Comment (simon-bar) [inline resources/rest-service/cloudify/policies/ewma_stabilized.clj]: change the name of the namespace

File: resources/rest-service/cloudify/types/types.yaml [hunk 15]
```diff
@@ -238,23 +238,49 @@ plugins:
...
-    cloudify.policies.types.host_failure:
-        source: https://raw.githubusercontent.com/cloudify-cosmo/cloudify-manager/master/resources/rest-service/cloudify/policies/host_failure.clj
+    cloudify.policies.types.ewma_stabilized:
+        properties:
+            <<: *THRESHOLD_BASED_POLICY_PROPERTIES
+            ewma_timeless_r:
+                description: |
+                    r is the ratio between successive events. The smaller it is, the smaller impact on the computed value the most recent event has.
+                default: 0.5
+        source: https://raw.githubusercontent.com/codilime/cloudify-manager/failure_healing_policies/resources/rest-service/cloudify/policies/ewma_stabilized.clj
```
Comment (simon-bar) [inline resources/rest-service/cloudify/types/types.yaml]: change after merge

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 16]
```diff
@@ -23,28 +31,48 @@
...
+                                                                  (get-policy-source-id event)
+                                                                  (sdo
+                                                                    {{stream.data}}
+                                                                    ;New line is needed so there can be a comment
+                                                                    ;in the last line of policy source
+                                                                  )))]
+  (where* (is-node-name-in {%for m in stream.metadata.members%} "{{m}}" {%endfor%})
+    (exception-stream publish-policy-error
+      (fn [event]
+        (def node-instances-stream
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: sdo

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 17]
```diff
@@ -23,28 +31,48 @@
...
+          (if (contains? (get-policy-source-location node-instances-stream) (get-policy-source-id event))
+            node-instances-stream
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: fix a bug

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 18]
```diff
@@ -23,28 +31,48 @@
...
+                                                   metadata-stream-id
+                                                   (:node_id event)))
+      add-policy-source-to-map     (fn [sources-map event] (assoc (get-policy-source-location sources-map)
+                                                                  (get-policy-source-id event)
+                                                                  (sdo
+                                                                    {{stream.data}}
+                                                                    ;New line is needed so there can be a comment
+                                                                    ;in the last line of policy source
+                                                                  )))]
+  (where* (is-node-name-in {%for m in stream.metadata.members%} "{{m}}" {%endfor%})
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: single-thread

File: plugins/riemann-controller/riemann_controller/resources/deployment.config.template [hunk 19]
```diff
@@ -23,28 +31,48 @@
...
-                               "{{stream.metadata.policy}}")
-      publish-log
-        (publish-log* "{{ctx.blueprint.id}}"
-                      "{{ctx.deployment.id}}"
-                      "{{stream.metadata.group}}"
-                      "{{stream.metadata.policy}}")]
-(where* (is-node-name-in {%for m in stream.metadata.members%}"{{m}}" {%endfor%})
-(exception-stream publish-policy-error
-{{stream.data}})))
+(let [ctx {:blueprint-id  "{{ctx.blueprint.id}}"
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/deployment.config.template]: check the underscores

File: plugins/riemann-controller/riemann_controller/resources/manager.config [hunk 20]
```diff
@@ -408,17 +403,80 @@
...
+  (let [node-id               (:node-id ctx)
+        manager-ip            (or (System/getenv "MANAGEMENT_IP") "127.0.0.1")
+        raw-manager-rest-port (or (System/getenv "MANAGER_REST_PORT") "80")
+        manager-rest-port     (Integer/parseInt raw-manager-rest-port)
+        base-uri              (str "http://" manager-ip ":" manager-rest-port)
+        node-endpoint         (str "/node-instances/" node-id)
+        node-resource-uri     (str base-uri node-endpoint)
+        get-node              (fn [] (clj-http.client/get node-resource-uri {:accept :json}))
+        get-state             (fn [node-response]
+                                (:state (cheshire.core/parse-string (:body node-response) true)))]
```
Comment (simon-bar) [inline plugins/riemann-controller/riemann_controller/resources/manager.config]: get-status and check if it is 404

**Pull request merged**